cardinal/server/validator/validator.go

```diff
@@ -0,0 +1,180 @@
+package validator
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/coocood/freecache"
+	"github.com/ethereum/go-ethereum/common" // for hash
+	"github.com/rotisserie/eris"
+
+	"pkg.world.dev/world-engine/sign"
+)
+
+type SignerAddressProvider interface {
+	GetSignerAddressForPersonaTag(personaTag string) (addr string, err error)
+}
+
+const cacheRetentionExtraSeconds = 10 // this is how many seconds past normal expiration a hash is left in the cache.
+// we want to ensure it's long enough that any message that's not expired but
+// still has its hash in the cache for replay protection. Setting it too long
+// would cause the cache to be bigger than necessary
+
+const ttlMaxFutureSeconds = 2 // this is how many seconds in the future a message is allowed to be timestamped
+// this allows for some degree of clock drift. It's safe enought to accept a message that's stamped from the near
+// future because we will still keep it in the hash cache and prevent it from being a replay attack vector. However,
+// we don't want to take messages from an unlimited amount of time into the future since they could cause our hash
+// cache to overflow
+
+const bytesPerKb = 1024
+
+var (
+	ErrNoPersonaTag     = errors.New("persona tag is required")
+	ErrWrongNamespace   = errors.New("incorrect namespace")
+	ErrMessageExpired   = errors.New("signature too old")
+	ErrBadTimestamp     = errors.New("invalid future timestamp")
+	ErrCacheReadFailed  = errors.New("cache read failed")
+	ErrCacheWriteFailed = errors.New("cache store failed")
+	ErrDuplicateMessage = errors.New("duplicate message")
+	ErrInvalidSignature = errors.New("invalid signature")
+)
+
+type SignatureValidator struct {
+	IsDisabled               bool
+	MessageExpirationSeconds int
+	HashCacheSizeKB          int
+	namespace                string
+	cache                    *freecache.Cache
+	signerAddressProvider    SignerAddressProvider
+}
+
+type ValidationError interface {
+	error
+	GetLogMessage() string
+	GetStatusCode() int
+}
+type validationError struct {
```

Comment:
This custom error type is not necessary and we generally try to avoid since the basic error primitive is more than sufficient.

1. I'd prefer not to control the status code from the validator because this breaks separation of concern; the validator doesn't necessarily have to be consumed by an HTTP server. Instead, in the server handler, use `errors.Is`/`eris.Is` to check for sentinel errors (the one you declared above) and handle the HTTP response accordingly there.

2. Instead of using `LogMsg`, you can just `eris.Wrap` your sentinel error to provide additional context to the error which you can log/print out.

By doing these 2 things, you should be able to eliminate this custom error type.